### ISSUE TITLE
More comprehensive test coverage

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#9493]
Updated Feb 23, 2015.

This would make a whole bunch of good "first contribution" projects, and does not require deep insider knowledge of the language.

The basic idea is to expand the test suite to make sure that julia's base code works as promised. Here is one recommended way to contribute toward this goal:

## The easy way

1. Go visit https://coveralls.io/r/JuliaLang/julia. (At the time of this writing, coverage numbers are in the 70s%. If you see numbers that are much lower than this, don't trust the reports; there is likely a problem in measuring coverage.)

2. Browse through the source files and find some untested functionality (highlighted in red) that you think you might be able to write a test for. 

3. Write a test that exercises this functionality---you can add your test to one of the existing files, or start a new one, whichever seems most appropriate to you. If you're adding a new test file, make sure you include it in the list of tests in `test/runtests.jl`. http://julia.readthedocs.org/en/latest/stdlib/test/ may be helpful in explaining how the testing infrastructure works. Submit the test as a pull request (see [CONTRIBUTING.md](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md)).

## The manual method

1. Make sure you have an up-to-date git checkout and are on the `master` branch. Build julia with `make`

2. Copy/paste the following script into an editor window and save it. Here I'll use the filename `/tmp/coverage_tests.jl`:
```julia
using Base.Test
testnames = ["core", "keywordargs", "numbers", "strings", "dates",
    "hashing", "remote", "iobuffer", "staged", "arrayops",
    "subarray", "reduce", "reducedim", "random", "intfuncs",
    "simdloop", "blas", "fft", "dsp", "sparse", "bitarray", "copy", "math",
    "functional", "bigint", "sorting", "statistics", "spawn",
    "backtrace", "priorityqueue", "arpack", "file", "suitesparse", "version",
    "pollfd", "mpfr", "broadcast", "complex", "socket",
    "floatapprox", "readdlm", "reflection", "regex", "float16", "combinatorics",
    "sysinfo", "rounding", "ranges", "mod2pi", "euler", "show",
    "lineedit", "replcompletions", "repl", "test", "goto",
    "llvmcall", "grisu", "nullable", "meta", "profile",
    "libgit2", "docs", "base64", "pkg", "linalg1", "linalg2",
	"linalg3", "linalg4", "linalg/lapack", "linalg/triangular", "linalg/tridiag", 
	"linalg/pinv", "linalg/cholmod", "linalg/umfpack", "linalg/givens"
]
for tst in testnames
    println(tst)
    include(joinpath(JULIA_HOME,Base.DATAROOTDIR,"julia","test","$tst.jl"))
end
```
(This is the list of tests currently in `test/runtests.jl`, with 3 omissions (`resolve`, `reflection`, and `meta`) and one addition (`pkg`). The omitted tests explicitly test inlining, which we're going to disable, or are problematic when inlining is disabled.)

3. From the top-level directory, type `rm usr/lib/julia/sys.so`. Deleting `sys.so` will prevent julia from using any pre-compiled functions, increasing the accuracy of the results. (This also makes startup a bit slower, but that's fine for this test---and once you're done, simply typing `make` will cause this file to be rebuilt).

4. Via the command prompt, navigate to julia's `test/` directory.

5. Start julia with `julia --code-coverage=all --inline=no`. This turns on code-coverage and prevents inlining (inlining makes it difficult to accurately assess whether a function has been tested).

6. At the julia prompt, type `include("/tmp/coverage_tests.jl")`.

7. Once the tests complete successfully, quit julia

8. Navigate to the `base/` directory.

9. Browse through the *.jl.cov files, and look for functions (or branches within functions) that either have `0` in front of them (indicating that they were run 0 times), or have `-` in front of them (indicating that they were never compiled).

10. Write a test that exercises some function that is not adequately covered---you can add your test to one of the existing files, or start a new one, whichever seems most appropriate to you. If you're adding a new test file, make sure you include it in the list of tests in `test/runtests.jl`. http://julia.readthedocs.org/en/latest/stdlib/test/ may be helpful in explaining how the testing infrastructure works. Submit the test as a pull request (see [CONTRIBUTING.md](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md)).

**Comment by timholy** [comment]
@svaksha, I remembered your `linalg4` issue; that gist you posted showed success, so clearly it's not "failing reliably." You might want to try a second time running the whole test suite as described above again.

If you experience the same `transpose(qrfact(...` error again, it would be great to open a separate issue.

**Comment by timholy** [comment]
OK, I've substantially edited the instructions above; with these changes, I _think_ we get reasonably-accurate results, as long as a human parses the `*.jl.cov` files.

However, I'm currently running into #9536, so (for me) currently this doesn't work.

**Comment by kshyatt** [comment]
@timholy I am having the same problem as @IainNZ. I tried your example with intersect, which worked perfectly. But when I follow the guide in the top post (including --inline=no) I'm getting dashes on lines which are definitely called. I'm on commit b06873f on OSX 10.10. I tried your fixed instructions in the top post (deleting sys.dylib) as well but I'm still having the problem.

**Comment by svaksha** [comment]
On Thu, Jan 1, 2015 at 3:15 PM, Tim Holy <notifications@github.com> wrote:
>
> OK, I've substantially edited the instructions above; with these changes, I think we get reasonably-accurate results, as long as a human parses the *.jl.cov files.
>
> However, I'm currently running into #9536, so (for me) currently this doesn't work.

Thanks for the update.

On Ubuntu-14.04, with commit
`3f3998e408aa1aec0af6d767ff7cf42ccf9c50fd`,  `make` failed with this
error after deleting  `rm usr/lib/julia/sys.so`

```
~/julia$ make
    CC src/jltypes.o
    CC src/gf.o
    CC src/support/hashing.o
    CC src/support/timefuncs.o
    CC src/support/ptrhash.o
    CC src/support/operators.o
    CC src/support/utf8.o
    CC src/support/ios.o
    CC src/support/htable.o
    CC src/support/bitvector.o
    CC src/support/int2str.o
    CC src/support/libsupportinit.o
    CC src/support/arraylist.o
    CC src/support/strtod.o
    LINK src/support/libsupport.a
    CC src/flisp/flisp.o
    CC src/flisp/builtins.o
    CC src/flisp/string.o
    CC src/flisp/equalhash.o
    CC src/flisp/table.o
    CC src/flisp/iostream.o
    CC src/flisp/julia_extensions.o
    LINK src/flisp/libflisp.a
    CC src/flisp/flmain.o
    LINK src/flisp/flisp
    FLISP src/julia_flisp.boot
    FLISP src/julia_flisp.boot.inc
    CC src/ast.o
    CC src/builtins.o
    CC src/module.o
    CC src/codegen.o
    CC src/disasm.o
    CC src/debuginfo.o
    CC src/interpreter.o
    CC src/alloc.o
    CC src/dlload.o
    CC src/sys.o
    CC src/init.o
    CC src/task.o
    CC src/array.o
    CC src/dump.o
    CC src/toplevel.o
    CC src/jl_uv.o
    CC src/jlapi.o
    CC src/profile.o
    CC src/llvm-simdloop.o
    CC src/gc.o
    LINK usr/lib/libjulia.so
/home/mom/julia/usr/lib/libunwind.a(Los-linux.o): In function
`_ULx86_local_resume':
Los-linux.c:(.text+0x75b): warning: sigreturn is not implemented and
will always fail
```
And, since yesterday,  @staticfloat's nightlies are not working -
strangely, `julia` (anywhere) and `./julia` (git master)  both show
the same version:
```
Version 0.4.0-dev+2396 (2015-01-02 07:26 UTC)
master/3f3998e (fork: 140 commits, 8 days)
```
Any idea what is the bug, if any?
tia,
SVAKSHA ॥  http://about.me/svaksha  ॥

**Comment by tkelman** [comment]
@svaksha it looks like that's a warning which you've gotten before? e.g. https://github.com/JuliaLang/julia/issues/7676 - if `make test` doesn't succeed, then please open a separate issue.

On test coverage, I just found that interestingly our tests all pass even without a `doc/helpdb.jl` file present, you get a `could not open file` error the first time you try calling help on something. That would seem to indicate help mode has no (or very few?) tests.

**Comment by timholy** [comment]
@kshyatt, glad that fixed it for you. @IainNZ, for you it was probably also the multiprocess issue. Sorry I was slow to realize that one needs to add JULIA_CPU_CORES=1 to that launch command.

**Comment by svaksha** [comment]
On Thu, Jan 1, 2015 at 1:25 PM, Tim Holy <notifications@github.com> wrote:
>
> @svaksha, I remembered your linalg4 issue; that gist you posted showed success, so clearly it's not "failing reliably." You might want to try a second time running the whole test suite as described above again.
>

@timholy, I am unable to reproduce the same error - ran it 3 times.
Seems like a fluke.

SVAKSHA ॥  http://about.me/svaksha  ॥

**Comment by svaksha** [comment]
On Fri, Jan 2, 2015 at 10:43 AM, Tony Kelman <notifications@github.com> wrote:
> @svaksha it looks like that's a warning which you've gotten before? e.g.
> #7676 - if make test doesn't succeed, then please open a separate issue.

@tkelman, thanks. #9555

**Comment by chrisvoncsefalvay** [comment]
`--inline=no` gives me an error:

```
ERROR: unknown option: --inline=no
 in process_options at /Applications/Julia-0.4.0-dev-d066c3f061.app/Contents/Resources/julia/lib/julia/sys.dylib
 in _start at /Applications/Julia-0.4.0-dev-d066c3f061.app/Contents/Resources/julia/lib/julia/sys.dylib
```

Runs just fine without it.

**Comment by timholy** [comment]
You have to be running a very recent julia for that to work.

**Comment by chrisvoncsefalvay** [comment]
I.e. more recent than `d066c3f`?

**Comment by garborg** [comment]
Yeah, [searching for '--inline'](https://github.com/JuliaLang/julia/search?utf8=%E2%9C%93&q=--inline&type=Code), looks like a commit @timholy got merged 3 days ago, and I believe your commit is just a hair older.

**Comment by hayd** [comment]
Just braindumping here: I think there is something to be said (along with coverage) for "organisation" of tests, that is to say locating tests in consistent files (since we don't have [cleverer ways](https://github.com/smangano/JLTest) to do it in base)... for the most part:

- foo defined in base/bar.jl 
- foo is tested in test/bar.jl

*(I would say this is often the case right now...)*

A rough heuristic of how well a function is tested might be to count how many times that function was called in test dir (or the specific test file)... you could also do the number of lines vs lines in its definition (or the calls on specific types if several methods are defined)...

I think this would give an interesting "hitlist" of functions to test better (in the above PR I just looked up those methods which aren't called *at all*), but it would be interesting to see what's the next "poorest covered" method to test next.

*I guess in someways this is independent of coverage but I thought I'd share my thought.*

**Comment by timholy** [comment]
Interesting ideas. Regarding organization of test files, I don't have strong feelings and will let others comments, but certainly there seems to be some logic to your proposal.

Regarding measurements of coverage, I think a lot of what you're proposing could be done by expanding the analysis that's done in Coverage.jl. You've obviously been reading through the `*.cov` files, so you know that line counts are already available. Coverage.jl now (thanks to @jakebolewski's excellent JuliaParser package) also now "understands" methods. So one could do some analysis on a method-by-method basis, sort methods by how many times they've been called, etc.

**Comment by ViralBShah** [comment]
I think it would be useful to capture what is discussed here in the documentation, along with some guidelines on writing tests. That will perhaps get a lot more people going. We can even advertise this on the website - as a great way to get started with julia.

**Comment by quinnj** [comment]
@hayd, FWIW, your suggestion is how the `Dates` module is organized (see `base/dates/` and `test/dates`). I agree that it was a very natural way to stay organized and improve coverage (with the recent coverage changes, Dates.jl coverage went from 99.73% to 83.42%, so I have a little more work to do!)

**Comment by nalimilan** [comment]
@quinnj Pretty impressive rates, still!

**Comment by hayd** [comment]
@quinnj and +1 for the Dates tests :)

I guess I was hoping to distinguish coverage between functions called "in anger" (i.e. in Base in a function which is tested) vs tests specifically against that method - looking for edge cases etc. (in a way you can only do if testing that method separately.

In python with unittest, for method foo, you might have a class TestFoo and all the tests are test_this_foo_thing methods... and you can quickly look over TestFoo, you may also want to check out the coverage from just running the TestFoo class. (i.e. for each function foo lookup it's test class and check its coverage).

IMO this flexibility is one of the benefits with a testing framework (that danluu was griping about in his blogpost)... I don't think you can do it with tests as scripts. :walking:

**Comment by StefanKarpinski** [comment]
It's somewhat embarrassing to admit, but a very long time ago there was a bug that made it problematic to have a base file and a test file with the same name, so we gave them different names intentionally :-\

**Comment by staticfloat** [comment]
It shocks me that Stefan can talk about things "a very long time ago" that I remember.

**Comment by IainNZ** [comment]
Re the test file - base file thing, I'm doing something like that in this PR: https://github.com/JuliaLang/julia/pull/9679

**Comment by chrisvoncsefalvay** [comment]
I'm trying to do this but with the current master, I get:
```
julia> include("/tmp/coverage_tests.jl")
linalg
ERROR: LoadError: could not open file /tmp/linalg.jl
 in include at ./boot.jl:248
 in include_from_node1 at ./loading.jl:128
 in anonymous at no file:17
 in include at ./boot.jl:248
 in include_from_node1 at ./loading.jl:128
while loading /tmp/coverage_tests.jl, in expression starting on line 15
```

Why's that? I feel like I'm missing something ;)

**Comment by tkelman** [comment]
try replacing `include("$tst.jl")` with `include(joinpath(JULIA_HOME,Base.DATAROOTDIR,"julia","test","$tst.jl"))`

**Comment by chrisvoncsefalvay** [comment]
I just checked, the problem seems to be that there is indeed no `linalg.jl`. Hmmm... bug or feature? :)

**Comment by tkelman** [comment]
Oh, right, linalg is a group of tests, it gets replaced by `["linalg1", "linalg2", "linalg3", "linalg4", "linalg/lapack", "linalg/triangular", "linalg/tridiag", "linalg/pinv", "linalg/cholmod", "linalg/umfpack", "linalg/givens"]` in that list

**Comment by chrisvoncsefalvay** [comment]
That's one problem solved ;) Now, however, it seems it cannot find `@test`:

```
ERROR: LoadError: LoadError: UndefVarError: @test not defined
 in include at ./boot.jl:248
 in include_from_node1 at ./loading.jl:128
 in anonymous at no file:18
 in include at ./boot.jl:248
 in include_from_node1 at ./loading.jl:128
 in eval at no file
while loading /Users/mywittyusername/Developer/julia/usr/bin/../share/julia/test/core.jl, in expression starting on line 7
while loading /tmp/coverage_tests.jl, in expression starting on line 16
```

**Comment by tkelman** [comment]
`using Base.Test` should fix that, or `include("testdefs.jl")`

**Comment by chrisvoncsefalvay** [comment]
Excellent. Worky working now. Thanks for shepherding this testing-noob ;)

@timholy can we upd the `coverage_tests.jl` you provided to the following:

```julia
using Base.Test
testnames = ["core", "keywordargs", "numbers", "strings", "dates", "collections", 
	"hashing", "remote", "iobuffer", "staged", "arrayops",
    "subarray", "reduce", "reducedim", "random", "intfuncs",
    "simdloop", "blas", "fft", "dsp", "sparse", "bitarray", "copy", "math",
    "functional", "bigint", "sorting", "statistics", "spawn",
    "backtrace", "priorityqueue", "arpack", "file", "suitesparse", "version",
    "pollfd", "mpfr", "broadcast", "complex", "socket",
    "floatapprox", "readdlm", "reflection", "regex", "float16", "combinatorics",
    "sysinfo", "rounding", "ranges", "mod2pi", "euler", "show",
    "lineedit", "replcompletions", "repl", "test", "goto",
    "llvmcall", "grisu", "nullable", "meta", "profile",
    "libgit2", "docs", "base64", "pkg", "linalg1", "linalg2",
	"linalg3", "linalg4", "linalg/lapack", "linalg/triangular", "linalg/tridiag", 
	"linalg/pinv", "linalg/cholmod", "linalg/umfpack", "linalg/givens"
]
for tst in testnames
    println(tst)
    include(joinpath(JULIA_HOME,Base.DATAROOTDIR,"julia","test","$tst.jl"))
end
```

**Comment by timholy** [comment]
@chrisvoncsefalvay, I need to update this page. There's now a much easier way: go visit https://coveralls.io/r/timholy/julia and start browsing through the files. It's built nightly (momentarily by me executing a script, but someday I'll figure out the bug in my cron script...).

**Comment by timholy** [comment]
Updated, see revised top post.

**Comment by chrisvoncsefalvay** [comment]
@timholy Ah-hah! Thank you ;)

**Comment by timholy** [comment]
Ah, thanks for the revision @chrisvoncsefalvay. Updated that, too.

**Comment by hayd** [comment]
@timholy Is [this](https://coveralls.io/files/422931268) a line highlighting bug?

Also should some lines (sparingly) by pragma no covered ? e.g. https://coveralls.io/files/422931392#L35 though I guess that specific example is about catching specific error types... #7026.

**Comment by timholy** [comment]
In your first example, which line are you referring to on that page? Number 37?

I've seen occasional oddities myself. One of the most likely explanations is #1334. Also see #9663. Any bug in line numbering is guaranteed to translate into bugs in coverage analysis.

**Comment by chrisvoncsefalvay** [comment]
Have now started covering (no pun intended) `Base.graphics` (just thought I'll put it here so we don't accidentally multiply our efforts - we could use a greenie board or a Trello checking who does what so we don't accidentally overlap!)

**Comment by timholy** [comment]
Good plan!

**Comment by chrisvoncsefalvay** [comment]
I [created a Trello](https://trello.com/b/7I8XZxIU/julia-coverage-board) for us to keep stuff coordinated. Happy to add any of you as users, [email me your username](mailto:chris@chrisvoncsefalvay.com).

**Comment by chrisvoncsefalvay** [comment]
@timholy Can you add adding the test to `tests/runtests.jl` to your original post, so that the new test would be run by the CI suites? Thanks ;)

**Comment by timholy** [comment]
Done.

**Comment by timholy** [comment]
Satisfying observation: coverage has gone up every single day since I started regularly [measuring it](https://coveralls.io/r/timholy/julia) (on Jan 11). The net increase is from 70.4% to 71.6%, but when you consider the size of julia's code base, that's pretty good. At this rate we'll hit 100% :smile: in about 7 months.

And those increases omit early test-coverage commits from @kshyatt and @hayd that occurred before I had set up the [necessary infrastructure](https://github.com/timholy/CoverageBase.jl).

**Comment by hayd** [comment]
@timholy one annoying thing with coveralls is that it reports 0/0 as 0% e.g. https://coveralls.io/files/433294481 so this will remain first in the sorted list of files.

I wonder if it's worth flagging one of those lines (perhaps module or include) so that it reports instead as 100% (yes, that'll inflate the numbers a little, but it'll also make it easier to find the next least covered file candidate).

80% by Spring. :)

**Comment by timholy** [comment]
We can't control how coveralls chooses to interpret 0/0; all we do is [submit a string containing the code and a single value per line](https://github.com/IainNZ/Coverage.jl/blob/8877c79a36787d25a67b16d21bf726b6183acbcc/src/Coverage.jl#L125-L127) containing either `null` (javascript analog of `nothing`) or the number of times that line ran.

One option is that we could just exclude files that have no runnable lines (see https://github.com/IainNZ/Coverage.jl/issues/45). As @garborg correctly pointed out, whatever we choose to do with files that have no runnable lines will have no impact on the final percentage.

Perhaps it would be worth asking them whether such files could be moved to the end of the list, and scored as NA rather than 0%?

**Comment by timholy** [comment]
Oh, I read your strategy more carefully; we could lie that way, I guess, but I like the idea of either excluding the files or contacting Coveralls.io about this better.

**Comment by garborg** [comment]
Contacting Coveralls would be great. (I was going to say I was for the change to NA and ambivalent about the move to the end, but for us, at least, it seems like a clear improvement, too.)

I do like keeping the 0/0 files in, though (second best option being doing nothing, imo). They're worth showing to people (at least leaving available, no one has to click on them) -- both because the tooling has TODOS (functions defined via macros, etc.) and may have regressions, and because those files are often filled with types, and the moment someone is scanning through Coveralls results is a moment it's likely to register that a subtype is left out completely or from a chunk of the tests.

**Comment by timholy** [comment]
https://github.com/lemurheavy/coveralls-public/issues/218

**Comment by dhoegh** [comment]
@timholy The changes I made in #9485 you merged more than a week ago does not seem to be included in the coverage result in https://coveralls.io/files/440827056. So I think there is something that does not seem to work with the daily updates of the results.

**Comment by timholy** [comment]
Indeed, sorry about that. I was debugging my cron script and commented out the `git pull; make cleanall && make` part. Oops. Can you check again?

**Comment by dhoegh** [comment]
https://github.com/JuliaLang/julia/blob/master/base/replutil.jl#L175 still seems to different from https://coveralls.io/files/440930474#L175

**Comment by dhoegh** [comment]
When you click at GitHub in coveralls it redirects you to your https://github.com/timholy/julia, which also seems to be behind.

**Comment by timholy** [comment]
Good debugging. I had convinced myself that it only used the source that we packaged up, but clearly it needs GitHub repo to be updated, too.

This will be a little hard to automate, because of the need to use the SSH key when pushing to https://github.com/timholy/julia; in the short term I'll try to remember to update it daily.

**Comment by dhoegh** [comment]
Thank you, the results also looks more correct now:) I can see that one code path I introduced in #9485 is untested, I thought I had covered it and but apparently not and I did not check it locally. I am going to submit a pull.

**Comment by brk00** [comment]
How can I check if a function has "raised" a Warning?

**Comment by timholy** [comment]
I'm not sure I understand what you're asking.

**Comment by johnmyleswhite** [comment]
Warnings aren't thrown (just printed), so you can't check them without monitoring STDOUT. You can check errors with `test_throws` because errors are thrown.

**Comment by brk00** [comment]
Thanks

**Comment by dhoegh** [comment]
The coveralls test has not run the last two days.

**Comment by quinnj** [comment]
I believe @timholy is running it manually at night, so let's not beat him
up too much :)

**Comment by staticfloat** [comment]
My automatic runs are held up by #10027

**Comment by timholy** [comment]
I appreciate the support @quinnj, but in this case @timholy frankly appreciates that (1) somebody is looking and (2) the same somebody is willing to ping him when something goes wrong.

I assumed you meant that it just wasn't running, so I ran a new one. But looking I now see that the coverage status is just borked. This is probably because of https://github.com/IainNZ/Coverage.jl/issues/48. Let me see what I can do about that.

**Comment by catawbasam** [comment]
On Coveralls TIMHOLY / JULIA / 41, clicking a file name to view source yields the following message:
```
SOURCE NOT AVAILABLE
The owner of this repo needs to re-authorize with Github; their OAuth credentials are no longer valid so the file cannot be pulled from the Github API.
```

**Comment by timholy** [comment]
We seem to be back in business now. Let me know of any further troubles.

**Comment by catawbasam** [comment]
Coveralls coverage % dropped roughly in half with JULIA / 49.
More accurate, or more trouble?

**Comment by timholy** [comment]
Presumably it's trouble. But it seems to be a base-julia problem.

**Comment by timholy** [comment]
Update: things are back to working with the local run, but as of today I'm having issues with [GnuTLS](https://github.com/JuliaWeb/GnuTLS.jl/pull/35).

**Comment by timholy** [comment]
And we're back in business. Up to 74%!

**Comment by sbromberger** [comment]
@timholy glad to hear. The GnuTLS failure was related to https://github.com/JuliaLang/julia/commit/2af73dd7723c91b0e9fcf5af7662b40c8fb97067.

**Comment by timholy** [comment]
Looks like @staticfloat has gotten the buildbots working! If you're feeling ready to make that the official version, OK if I update the link above?

(Looks like my days of running things manually can thankfully come to a close....)

**Comment by staticfloat** [comment]
I still haven't figured out how to get the Git information to propagate properly, but other than that things seem to be working just fine.

**Comment by brk00** [comment]
Coveralls isnt's working since march 7th. Anyone knows why?

**Comment by staticfloat** [comment]
There have been a couple bugs that killed things, one of which was https://github.com/jakebolewski/JuliaParser.jl/issues/18 which was just fixed.  The latest failure is due to [a segfault](http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64/builds/309/steps/Run%20non-inlined%20tests/logs/stdio).

**Comment by kshyatt** [comment]
I looked at Coveralls today and it seems to be behaving weirdly. For example, let's look at https://coveralls.io/builds/2460415/source?filename=base%2Fstatistics.jl. Coveralls claims this has 0% test coverage for the latest build. Yet, when I actually look in `test/`, I see that `test/statistics.jl` is there and has lots of great tests. There are a LOT of files that Coveralls says have no test coverage but have files full of tests in `test/`. Are they just not being run? It's a bit frustrating for those who want to write more tests if the coverage results are totally wrong.

**Comment by hayd** [comment]
Has the coverage of Julia dropped to 26%?? https://coveralls.io/r/JuliaLang/julia

(The coverall's page shows 73% at the top level, but 26% on each build.)

**Comment by ihnorton** [comment]
No; see https://github.com/JuliaLang/julia/issues/10804

**Comment by ihnorton** [comment]
Now fixed. I'm not sure what magic wand must be waved to regenerate those numbers, but presumably they should go up if @staticfloat's report was the only problem.

**Comment by timholy** [comment]
Thanks, @ihnorton!

**Comment by tkelman** [comment]
Better, but still some bugs apparently http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64

**Comment by kshyatt** [comment]
Some more Coveralls builds ran last night and the file coverage stats are still bad. Is there a backlog, or more bug-hunting to do?

**Comment by tkelman** [comment]
More bug-hunting. Reflection http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64/builds/397/steps/Run%20non-inlined%20tests/logs/stdio, and linalg/symmetric http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64/builds/397/steps/Run%20inlined%20tests/logs/stdio have issues in the buildbot job that's uploading to coveralls

**Comment by kshyatt** [comment]
I know less than nothing about how the buildbots work with Coveralls. Are the failures only on Ubuntu? Isn't the Travis build succeeding for Linux (is it a different Linux)? This is probably an **extremely** stupid question, but can Coveralls just scrape the currently working Travis tests - are its buildbots just or something?

**Comment by tkelman** [comment]
Not a stupid question at all, the buildbot stuff is unfortunately opaque unless you actively bug staticfloat (or me, now, I suppose) until you figure out how it works :). We're only trying to run automated base coverage for regular coveralls uploads from an Ubuntu VM right now, yes. The commands that it runs are configured here https://github.com/staticfloat/julia-buildbot/blob/master/master/coverage.py

I believe the reason we can't do coverage results for base from Travis is it requires running without inlining, and I think on a single test worker, to get the most accurate results. So we would run out of time if we tried to do it on Travis, and running it for every single PR build would probably be disruptive at this point. Might be worth aiming for long-term though.

**Comment by staticfloat** [comment]
@tkelman is right on all points.  A typical run of the coverage stats takes ~10 minutes, which is a not-insignificant fraction of the amount of time we spend on Travis.  Additionally, we're only running the coverage stats on commits which pass tests on _all_ platforms, which is either a good thing or a bad thing, depending on how you look at it.

**Comment by kshyatt** [comment]
Thanks for the explanations! I don't have a machine running Linux handy but would it be worthwhile to see if I can reproduce the failures on OSX?

**Comment by staticfloat** [comment]
That would be handy, yes!  It's also helpful having someone helping to push at an issue, even if they aren't familiar with how to fix stuff, as it helps keep the issue closer to the "front burner" for people like myself and Tony, as it were. :)

**Comment by kshyatt** [comment]
OSX run in progress. Does the `reflection.jl` bug happen when inlining is turned off, or only when it's on?

**Comment by staticfloat** [comment]
We tend to run into bugs at this stage only when inlining is turned off, as the Travis and other buildbot checks are all using default settings (e.g. with inlining turned on).

**Comment by kshyatt** [comment]
BOOM! :boom: The `linalg/symmetric` bug is reproducible on OSX. I'll try to hunt this down today.

**Comment by tkelman** [comment]
Let's open a new issue for it so we can track it separately.

**Comment by kshyatt** [comment]
Done at #11089

**Comment by kshyatt** [comment]
I can also reproduce the `reflection.jl` bug on OSX (when run with no inlining). That bug apparently goes away **without** inlining since I'm running on one worker and don't hit it immediately. When I run with inlining, the tests immediately fail at `reflection.jl`. Shall I open a second issue for that?

**Comment by tkelman** [comment]
yesplz

**Comment by kshyatt** [comment]
#11092

**Comment by kshyatt** [comment]
Ok, I ran CoverageBase myself and you can check it out at https://coveralls.io/r/kshyatt/julia. The coverage stats are at 59% right now. It seems somewhat more accurate on a per-file basis.

**Comment by hayd** [comment]
@kshyatt I don't think this can't possibly be correct e.g. https://coveralls.io/builds/2471016/source?filename=base%2Fmarkdown%2FGitHub%2Ftable.jl#L53 (how can this be a hit, when the entire function doesn't appear to be called.) There's some other weirdness too. :S

**Comment by kshyatt** [comment]
Oh no! This is very sad. So both I and the main Juila Coveralls are wrong in terrible ways? Time for more bug-hunting I guess. Thanks for pointing that out @hayd!

**Comment by hayd** [comment]
(IMO it's a good thing... coverage of 59% would be a much worse thing that a bug in coverage! Fingers' crossed the actual coverage is higher! To me this suggests a need for better test coverage for test coverage... :suspect:)

**Comment by timholy** [comment]
The stats were around 74% when things were working properly. It's hard to credit a plunge of that magnitude. Are you using https://github.com/timholy/CoverageBase.jl?

I don't want to test right now because I'm in the middle of  some disruptive changes, but last I checked the reflection tests literally required inlining: https://github.com/timholy/CoverageBase.jl/blob/28749b2c2e93d9de7535548e279921a5a329a5e9/src/CoverageBase.jl#L5. If I remember correctly, the `code_warntype` tests rely on inlining [these functions](https://github.com/JuliaLang/julia/blob/aa756622ea349cd28788509f32b23a05a42dbbb4/test/reflection.jl#L50-L51).

**Comment by timholy** [comment]
Also re possible bugs: if I remember correctly, coveralls uses the statistics of what we send to it, but grabs the code from github (despite the fact that we also send the code). So if the version of julia on github is from what's being tested, then there may be mis-registration between the code and the stats.

**Comment by timholy** [comment]
OK, now I'm catching up with the deluge, and you guys are worried about a different reflection issue than the one I assumed. Carry on!

**Comment by kshyatt** [comment]
I'm using a new, unmodified fork of https://github.com/timholy/CoverageBase.jl. I followed the steps in the README but I guess I must have done something wrong (or the tests are dying somewhere)?

**Comment by timholy** [comment]
Tests dying seems most likely. Any error messages, either in running the tests or gathering the statistics?

**Comment by kshyatt** [comment]
I get a boatload of deprecation warnings, but no test errors. This is after the merge of the commits fixing the issues in `linalg/symmetric` and `reflection`.

**Comment by tkelman** [comment]
On buildbot it appears to be timing out, possibly freezing? http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64/builds/402/steps/Run%20non-inlined%20tests/logs/stdio

**Comment by kshyatt** [comment]
I think we have 2 separate problems here. On my Coveralls, I show a coverage % of 59%, which is far too low, but my tests are running and finishing successfully. On the main Julia Coveralls, the latest build has a coverage of 10% (!!!) possibly related to the buildbot dying. When I run the coverage on my local machine, it's spending *ages* in `ranges` and `linalg/pinv`. For some reason the `ranges` tests just take forever when run with coverage stats.

**Comment by tkelman** [comment]
> For some reason the `ranges` tests just take forever when run with coverage stats.

How long are we talking? It's probably just an indicator of how heavily some parts of Julia's standard library rely on inlining to get good performance. The timeout on the buildbot should be configurable somewhere.

**Comment by kshyatt** [comment]
I'm running on a 2012 Macbook Air and it can take upwards of half an hour for that specific set of tests.

**Comment by kshyatt** [comment]
I think what's happening is that the `ranges` tests produce no output for > 10 minutes, and Travis assumes the job is hanging and kills it - see [here](http://docs.travis-ci.com/user/build-timeouts/). We can get around this by prepending the `ranges` tests with `travis-wait`, which would give `ranges` 20 minutes to finish running. I don't know if that's enough time.

**Comment by tkelman** [comment]
The buildbots aren't running on Travis, they just appear to be configured with a default timeout that's a little short. If we can find the configuration option for that we can certainly change it.

**Comment by staticfloat** [comment]
Bumped in https://github.com/staticfloat/julia-buildbot/commit/386912661ac47fc5efb6eae0f13700a8da32e323.  I'll re-run the latest coverage to see if it still fails here.

**Comment by timholy** [comment]
It's surprising that a test should take this long. I looked into it, and it basically hangs [in this loop](https://github.com/JuliaLang/julia/blob/da0dc58a44b7caed160e69c656f31d6fd5ddb9a4/test/ranges.jl#L283-L301). It took me a while to read that correctly: note the comma at the end of line 283, which means it's a continuation of the `for` statement and hence a nest of 5 loops. When inlining is off, this become prohibitive.

CC @StefanKarpinski, who last touched those lines. Do we need all those test cases? It's a total of 790500 calls to `@test`, which without inlining seems to be slow. Changing them to `@assert` cut the time by about a factor of 3 under normal conditions, but that's not enough to make the tests complete in a reasonable amount of time with `--inline=no` (still waiting after 3 minutes). I also tried putting them into a function, and still no dice.

Profiling indicates that a lot of the time is in `cat`-related functionality---do you need to perform concatenation for those tests, or is it just a convenience? With inlining off and putting those into a function, with a very unscientific statistical profiling I obtained
```jl
julia> @time range_looptests()
^CERROR: InterruptException:
 in setindex! at array.jl:395
 in cat_t at abstractarray.jl:719
 in typed_vcat at abstractarray.jl:740
 in range_looptests at none:8
 in eval at no file

julia> @time range_looptests()
^CERROR: InterruptException:
 in call at base.jl:291
 in vcat at range.jl:619
 in range_looptests at none:16
 in eval at no file

julia> @time range_looptests()
^CERROR: InterruptException:
 in call at base.jl:38
 in length at range.jl:269
 in done at range.jl:324
 in vcat at range.jl:623
 in range_looptests at none:10
 in eval at no file

julia> @time range_looptests()
^CERROR: InterruptException:
 in call at base.jl:38
 in length at range.jl:269
 in vcat at range.jl:617
 in range_looptests at none:14
 in eval at no file

julia> @time range_looptests()
^CERROR: InterruptException:
 in next at ./tuple.jl:17
 in vcat at range.jl:616
 in range_looptests at none:15
 in eval at no file
```
i.e., 5/5 were in `cat` functionality (4/5 in `vcat`, 1/5 in `typed_vcat`).

**Comment by StefanKarpinski** [comment]
We could cut the number of tests down, but that was *very* useful in figuring out if FloatRange was working or not. I'm not sure what the most effective way to pick a nasty subset is...

**Comment by timholy** [comment]
I had the same concern about #10505. Do you have to call `cat`, or could we devectorize the tests?

**Comment by kshyatt** [comment]
Update: although the results haven't appeared on Coveralls yet, the last two coverage runs appear to have succeeded based on buildbot logs

**Comment by StefanKarpinski** [comment]
:tada:

**Comment by timholy** [comment]
@StefanKarpinski, to clarify: what I meant was, is the call to `cat` just a convenience or is it something you're explicitly testing here?

**Comment by StefanKarpinski** [comment]
What call to `cat`?

**Comment by quinnj** [comment]
:telephone_receiver: :smirk_cat:

**Comment by kshyatt** [comment]
Nothing in life is ever easy. I've been having a Coveralls error all day, where the tests will finish successfully but I get an error right at the end, 
`ASCIIString "{\"message\":\"Couldn't find a repository matching this job.\",\"error\":true}"`
I just checked the buildbot logs and it had the same problem, which explains why the main Julia Coveralls hasn't updated. :unamused:

**Comment by timholy** [comment]
@StefanKarpinski, [these lines](https://github.com/JuliaLang/julia/blob/da0dc58a44b7caed160e69c656f31d6fd5ddb9a4/test/ranges.jl#L295-L300) (plus 289 and 291).

**Comment by timholy** [comment]
@kshyatt, just checking: you're running the script like this, right?
```bash
REPO_TOKEN=uRv4Kxxxxxxxxxxxxx /usr/local/julia/packages/v0.4/CoverageBase/run_coverage.sh
```
where the REPO_TOKEN is taken from a Coveralls repository to which you have push access, right? (Don't publish your full REPO_TOKEN here unless you want other users to be able to write information to your Coveralls repo page :smile:.)

**Comment by kshyatt** [comment]
Yes, I am. If you check the [log](http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64/builds/423/steps/Merge%20and%20submit/logs/stdio) for the latest Julia coverage build on the buildbot it's got the same error.

**Comment by timholy** [comment]
I'm not sure what's happening---my understanding of these webby thingies is not very sophisticated. Can you (or @staticfloat?) check whether the repo token has changed?

**Comment by kshyatt** [comment]
I checked on my version - I regenerated the token, got a new one, and tried again. I think the problem is on Coveralls' end.

**Comment by timholy** [comment]
Hmm, looking at the relevant lines for [`vcat`](https://github.com/JuliaLang/julia/blob/900cf12409d57a21fd1d92263a0633edd1603e9d/base/range.jl#L614-L626) and comparing against those Ctrl-C interrups, I'm not sure that avoiding the allocation will help that much; 4/5 of those breaks occurred in lines that were looping, not creating arrays.

As usual, profiling was informative and turned up a surprising answer: the single biggest culprit is [a particular call to `length`](https://github.com/JuliaLang/julia/blob/900cf12409d57a21fd1d92263a0633edd1603e9d/base/range.jl#L269) via it's call to [`call`](https://github.com/JuliaLang/julia/blob/900cf12409d57a21fd1d92263a0633edd1603e9d/base/base.jl#L38). Notice that's a splatting function. After I fixed that, the next culprit was a call to [`typed_vcat`](https://github.com/JuliaLang/julia/blob/da0dc58a44b7caed160e69c656f31d6fd5ddb9a4/base/abstractarray.jl#L740), also splatting.

For me, #11158 reduced the run time of a simplified version of that loop-nest from 1.88 seconds to 0.1 seconds. I'd call that victory, although it's worth noting that with inlining off the ranges test still takes nearly 2 minutes on my machine. (It still might be good to try to cut those down a bit, although I think the main pressure to do so is gone.)

**Comment by IainNZ** [comment]
(FYI, I'm looking at adding codecov.io support to Coverage.jl, Coveralls.io seems awfully broken/slow a lot of the time)

**Comment by timholy** [comment]
@IainNZ, :+1:.

**Comment by staticfloat** [comment]
Plus they have a very nice landing page, which counts for a lot!
-E

**Comment by kshyatt** [comment]
codecov.io looks super cool but apparently doesn't have Julia as an officially supported language yet. Worth firing an email off to hello@codecov.io ?

**Comment by IainNZ** [comment]
I'm already working with them, but it doesn't matter that too much that it isn't official. Coveralls.io doesn't either, but they have a generic API so I just wrote code to fire JSON at them.

**Comment by kshyatt** [comment]
Coveralls is back, with a cool **75%** coverage! :balloon:

**Comment by hayd** [comment]
Are there tests for coverage? i.e. do you have an example julia file, run coverage on it without inlining and compare the cov output to expected?

**Comment by timholy** [comment]
@kshyatt, I think Coveralls got nervous about @IainNZ supporting other sites, and got their act together.

@hayd, AFAIK the best we have is https://github.com/JuliaLang/julia/blob/de6e41fae8a71fd94f06d1462ad1bf2d4a9b31c0/test/pkg.jl#L85-L125.

**Comment by kshyatt** [comment]
@staticfloat The build is still occasionally timing out and giving us bad coverage results. Should the time limit on the buildbot be bumped again, or should we get @timholy's `ranges` fix to pass tests and merge that? Or both?

**Comment by timholy** [comment]
It was failing just because of an inability to checkout METADATA. I pushed a new commit that just added comments, and all the tests passed, so you should be good to go now.

**Comment by kshyatt** [comment]
Awesome! Sorry I was such a :beetle: about this.

**Comment by tkelman** [comment]
@kshyatt if having up-to-date coverage reports helps facilitate the PR's you've been putting together, then we're the ones who owe you an apology for not keeping more on top of it. Please keep them coming, poke at more things until they break so we know more specifically what needs fixing.

**Comment by kshyatt** [comment]
@tkelman It helps to be able to say "what part of `base/linag` needs more coverage?" to prioritize easy things to test. Also, I have a huge weakness for "if your numbers go up, it means you're having more fun!"

**Comment by timholy** [comment]
That's the kind of weakness I'm quite happy to help support.

I'm in the middle of another fix or I'd test it myself, so let me just ask: how bad is the situation with the `linalg/pinv` tests? Does this need the same treatment?

**Comment by kshyatt** [comment]
On my MBA,
```julia
julia> @time CoverageBase.runtests(["linalg/pinv"])
linalg/pinv
elapsed time: 211.548127765 seconds (37641 MB allocated, 2.33% gc time in 1718 pauses with 1 full sweep)
```

**Comment by timholy** [comment]
That is indeed significantly worse. Unfortunately, in this case there doesn't seem to be a straightforward fix: the bottleneck (over 95% of the runtime) is triggered by the calls to `vecnorm` in `test_pinv`, and ultimately the actual bottlenecks are two `next` calls ([here](https://github.com/JuliaLang/julia/blob/5ef01ac44c09253f822c045de7d719e0d4a28283/base/abstractarray.jl#L370) and [here](https://github.com/JuliaLang/julia/blob/5ef01ac44c09253f822c045de7d719e0d4a28283/base/tuple.jl#L21)) which, without inlining, allocate new tuples on each iteration.

Presumably the only fix is to use smaller vectors/matrices. CCing @andreasnoack to see if that's worth it (it may not be, if things like roundoff error are part of the point of these tests). So we may just have to live with the slowness here.

**Comment by staticfloat** [comment]
3 minutes of slowness is not a problem.  An hour of slowness is.
-E

**Comment by timholy** [comment]
My feeling, too.

**Comment by andreasnoack** [comment]
The tests `pinv.jl` use relatively larger matrices that the rest of the linear algebra tests, but it might be that the author, @zgimbutas, had good reasons for this. These tests are not just about hitting all branches of the code, but to carefully check that the numerical error has the expected magnitude. We might want to split the tests into two groups for this reason.

**Comment by timholy** [comment]
Unless @zgimbutas tells us otherwise, I think the marginal benefit from using smaller matrices is not worth the risk.

**Comment by zgimbutas** [comment]
The tests are constructed to test the pseudo-inverses of singular matrices. I tried to use matrices that have sufficiently fast decaying spectra so the matrix dimensions can be reduced somewhat but I would be careful.

**Comment by IainNZ** [comment]
BTW, Codecov does seem to work if you use this PR:
https://github.com/IainNZ/Coverage.jl/pull/58
e.g.
https://codecov.io/github/IainNZ/Coverage.jl?ref=ird/codecov

What is the workflow used for generating coverage reports on for JuliaLang/julia? Its not running on Travis right?

**Comment by staticfloat** [comment]
> Its not running on Travis right?

Correct, it's running on the buildbots.  The script is [here](https://github.com/staticfloat/julia-buildbot/blob/master/master/coverage.py).  It basically uses `CoverageBase` to run whatever tests we can non-inlined and uploads them via `Coverage`.

**Comment by kshyatt** [comment]
Now up to **76%** coverage! Many files in `pkg/` are completely untested - is there some way we could make a "dummy" package to test them? I don't know much about the internals of this.

**Comment by StefanKarpinski** [comment]
There's a `test-pkg` target that doesn't get run by default because it does a bunch of unusual nonsense like cloning METADATA, which won't work everywhere. I guess what could work is to have a dummy METADATA repo and install a dummy package from that. It seems like a lot of stuff to check into the language repo, but then again, this is important functionality and it has been broken now and then.

**Comment by tkelman** [comment]
Might just need to add `pkg` to the output of `CoverageBase.testnames()`

**Comment by staticfloat** [comment]
> Might just need to add pkg to the output of CoverageBase.testnames()

Yeah, let's do that.

**Comment by staticfloat** [comment]
https://github.com/timholy/CoverageBase.jl/issues/4

**Comment by timholy** [comment]
@kshyatt, wow, that is fast improvement! I wonder who might be responsible for that progress...

@staticfloat, saw this after I saw your issue posted. I think it probably does make sense to add it to CoverageBase.

**Comment by kshyatt** [comment]
>I wonder who might be responsible for that progress...
:coffee:

**Comment by tkelman** [comment]
> E: Figured out why my coverage % was so bad. It's because I was running on OSX.

What did that change? That sounds unexpected to me.

**Comment by kshyatt** [comment]
I think the `parallel` tests are Linux-only.

**Comment by ViralBShah** [comment]
When do we close this issue? A yellow coverage badge?

**Comment by kshyatt** [comment]
I would personally prefer a green badge, but I'm happy to defer to others. There are areas (the REPL, especially) that have overall poor test coverage so maybe we could open individual issues for those?

**Comment by timholy** [comment]
I think this is one of those issues that's counterproductive to close until we hit 100%. After all, it is a great entry point for new developers, and we don't have that many good examples of that kind of issue.

I'll be super-excited to see that color change to yellow!

**Comment by ivarne** [comment]
> I think this is one of those issues that's counterproductive to close until we hit 100%

I agree, but when the number of comments becomes becomes high, it's probably time to archive the discussion and continue the work in a new issue with an updated summary on top.

**Comment by ViralBShah** [comment]
I agree. For other areas, it would be nice to have specific issues to direct people to write tests for those, and close the umbrella issue.

**Comment by kshyatt** [comment]
I'd like to have a few separate issues like "Improve REPL tests" or "Flesh out LAPACK tests" and keep the information in the OP here as a Gist/README somewhere. It's probably pretty intimidating for some people to go to the "more tests" issue and see 150+ comments.

**Comment by timholy** [comment]
If people prefer that, it's all fine by me.

**Comment by ViralBShah** [comment]
@kshyatt Great idea to add the Coveralls badge to the README. Go for it!

**Comment by kshyatt** [comment]
It's me being a pain again - Coveralls hasn't run in 4 days. Are the buildbots still down?

**Comment by staticfloat** [comment]
Thank you for the ping, I've restarted the latest builds after clearing out the build directory following multiple failures in a row over the last few days.  We'll see if things shake themselves out.

**Comment by staticfloat** [comment]
Things did not shake themselves out, I had to rebuild the CentOS 5 buildbots.  The good news is, that's almost completely automated now. We've got a new coverage build up, but alas, still no git info making up there.

@kshyatt I seem to be having difficulty communicating git commit information to Coveralls.  My mini script for submitting coverage results is:
```julia
r1 = load("coverage_noninlined.jld", "results")
r2 = load("coverage_inlined.jld", "results")
r = CoverageBase.merge_coverage(r1, r2)
git_info = @compat Dict(
    "branch" => Base.GIT_VERSION_INFO.branch,
    "remotes" => [
        @compat Dict(
            "name" => "origin",
            "url" => "https://github.com/JuliaLang/julia.git"
        )
    ],
    "head" => @compat Dict(
        "id" => Base.GIT_VERSION_INFO.commit,
        "message" => "Merge pull request #11390 from ScottPJones/spj/deputf32"
    )
)
println("git_info: ")
println(git_info)
Coveralls.submit_token(r, git_info)
```

Which results in: (You can also look at [the actual execution log](http://buildbot.e.ip.saba.us:8010/builders/coverage_ubuntu14.04-x64/builds/480/steps/Merge%20and%20submit/logs/stdio))
```
WARNING: char(x) is deprecated, use Char(x) instead.
 in depwarn at deprecated.jl:44
 in char at deprecated.jl:31
 in include at boot.jl:252
 in include_from_node1 at loading.jl:134
 in include at boot.jl:252
 in include_from_node1 at loading.jl:134
 in reload_path at loading.jl:158
 in _require at loading.jl:70
 in require at loading.jl:56
 in include at boot.jl:252
 in include_from_node1 at loading.jl:134
 in reload_path at loading.jl:158
 in _require at loading.jl:70
 in require at loading.jl:53
 in process_options at client.jl:288
 in _start at client.jl:409
while loading /home/ubuntu/.julia/v0.4/JuliaParser/src/lexer.jl, in expression starting on line 12
Warning: could not import Base.put into Requests
Coverage.MallocInfo[]
git_info: 
Dict{ASCIIString,Any}("branch"=>"master","head"=>Dict("message"=>"Merge pull request #11390 from ScottPJones/spj/deputf32","id"=>"04051e96993cb53c8556691ee40e49bd054c7a5b"),"remotes"=>[Dict("name"=>"origin","url"=>"https://github.com/JuliaLang/julia.git")])
WARNING: [a,b,...] concatenation is deprecated; use [a;b;...] instead
 in depwarn at deprecated.jl:44
 in oldstyle_vcat_warning at abstractarray.jl:28
 in vect at abstractarray.jl:37
 in render at /home/ubuntu/.julia/v0.4/Requests/src/Requests.jl:23
 in open_stream at /home/ubuntu/.julia/v0.4/Requests/src/Requests.jl:224
 in send_multipart at /home/ubuntu/.julia/v0.4/Requests/src/Requests.jl:519
 in do_request at /home/ubuntu/.julia/v0.4/Requests/src/Requests.jl:566
 in post at /home/ubuntu/.julia/v0.4/Requests/src/Requests.jl:583
 in submit_token at /home/ubuntu/.julia/v0.4/Coverage/src/Coverage.jl:236
 in process_options at client.jl:288
 in _start at client.jl:409
while loading no file, in expression starting on line 0
WARNING: int(x) is deprecated, use Int(x) instead.
 in depwarn at deprecated.jl:44
 in int at deprecated.jl:31
 in on_header_value at /home/ubuntu/.julia/v0.4/Requests/src/Requests.jl:122
while loading no file, in expression starting on line 0
ASCIIString "{\"message\":\"Job #15453.1\",\"url\":\"https://coveralls.io/jobs/6174181\"}"
```

Are you doing anything differently when you run things locally in order to generate the `git` information?  What is the payload you submit to `Coveralls` in order to get it to accept the `git` information?

**Comment by kshyatt** [comment]
I'm just using CoverageBase.jl. It's whatever method that package uses, which is presumably the same thing that's running on the buildbots! I've run coverage on julia.mit.edu and had it work - something might be weird about the buildbot?

**Comment by staticfloat** [comment]
Can you show me the results of your coverage run?

**Comment by ViralBShah** [comment]
I think we should close this issue - and have new issues for some of these discussions.

**Comment by samuelpowell** [comment]
A number of more general purpose parts of base are covered only by virtue of their use in other tests in the suite.

Is it fair to say that each module should be tested fully by its *own* test script, such that the modification of unrelated tests does not have the unintended consequence of leaving untested code?

**Comment by timholy** [comment]
I find it helpful to have tests that fail for obvious reasons rather than for indirect reasons. For example, if there's a problem with type inference, it's much nicer if it fails a direct test in `core.jl` than having it caught because some call dispatched to the wrong method. The second is much harder to deduce at the level of what went wrong and what needs changing.

**Comment by ScottPJones** [comment]
:+1: @samuelpowell @timholy Definitely saves a lot of time to have pinpoint tests.

**Comment by samuelpowell** [comment]
Excellent, I'll see what I can do.

**Comment by tkelman** [comment]
Let's close this in favor of https://github.com/JuliaLang/julia/issues/11885 and continued targeted pushes. The instructions in the top post are still valuable, but the nearly-200 comments discussion is not so relevant going forward and takes a while to load.
